assets/javascripts/lib/uploads.js

```diff
@@ -47,3 +54,37 @@ export function generateUploadKey() {
       .then(resolve, reject);
   });
 }
+
+export function downloadEncryptedFile(url, keyPromise, opts) {
+  opts = opts || {};
+
+  const downloadPromise = new Promise((resolve, reject) => {
+    let req = new XMLHttpRequest();
+    req.open("GET", url, true);
+    req.responseType = "arraybuffer";
+    req.onload = function () {
+      let filename = req.getResponseHeader("Content-Disposition");
+      if (filename) {
+        // Requires Access-Control-Expose-Headers: Content-Disposition.
+        filename = filename.match(/filename="(.*?)"/)[1];
+      }
+      resolve({ buffer: req.response, filename });
+    };
+    req.onerror = reject;
+    req.send(null);
+  });
+
+  return Promise.all([keyPromise, downloadPromise]).then(([key, download]) => {
+    const iv = download.buffer.slice(0, 12);
+    const content = download.buffer.slice(12);
+
+    return new Promise((resolve, reject) => {
```

Comment:
I'm curious about the need to do `new Promise` here when `window.crypto.subtle.decrypt` seems to return a promise? Could you not return that?

Reply:
I can, but the tests might not work anymore. The runloop is aware of Ember Promises, but is not of the native ones:

https://discuss.emberjs.com/t/readers-questions-why-does-ember-still-use-rsvp/14736